src/internal/forward_diff.jl

```diff
@@ -2,16 +2,33 @@
 import SimpleNonlinearSolve: __nlsolve_ad, __nlsolve_dual_soln, __nlsolve_∂f_∂p,
                              __nlsolve_∂f_∂u
 
-function SciMLBase.solve(
-        prob::NonlinearProblem{<:Union{Number, <:AbstractArray}, iip,
-            <:Union{<:Dual{T, V, P}, <:AbstractArray{<:Dual{T, V, P}}}},
-        alg::Union{Nothing, AbstractNonlinearAlgorithm},
-        args...;
-        kwargs...) where {T, V, P, iip}
-    sol, partials = __nlsolve_ad(prob, alg, args...; kwargs...)
-    dual_soln = __nlsolve_dual_soln(sol.u, partials, prob.p)
-    return SciMLBase.build_solution(
-        prob, alg, dual_soln, sol.resid; sol.retcode, sol.stats, sol.original)
+for (uType, pType) in [
```

Comment:
Why is this necessary? This seems pretty ugly

Reply:
ambiguity prevention

Reply:
is there a reason this doesn't overload `__init` and `__solve`  like `OrdinaryDiffEq does?

Reply:
I was asking myself the same question today morning

Reply:
i think you need to so that you promote first.